mlflow/types/schema.py

```diff
@@ -122,34 +125,280 @@ def from_numpy_type(cls, np_type):
+    ) -> None:
+        try:
+            self._dtype = DataType[dtype] if isinstance(dtype, str) else dtype
+        except KeyError:
+            raise MlflowException(
+                f"Unsupported type '{dtype}', expected instance of DataType, Array, Object or "
+                f"one of {[t.name for t in DataType]}"
+            )
+        if not isinstance(self.dtype, (DataType, Object)):
+            raise MlflowException(
+                "Expected mlflow.types.schema.Datatype, mlflow.types.schema.Object"
+                f"or str for the 'dtype' argument, but got {self.dtype.__class__}"
+            )
+
+    @property
+    def dtype(self) -> Union[DataType, ObjectType]:
+        """The array data type."""
+        return self._dtype
+
+    def __eq__(self, other) -> bool:
+        if isinstance(other, Array):
+            return self.dtype == other.dtype
+        return False
+
+    def to_dict(self):
+        items = self.dtype.name if isinstance(self.dtype, DataType) else self.dtype.to_dict()
+        return {"type": "array", "items": items}
+
+    @classmethod
+    def from_json_dict(cls, **kwargs):
+        """
+        Deserialize from a json loaded dictionary.
+        The dictionary is expected to contain `type` and
+        `items` keys.
+        """
+        if not {"items", "type"} <= set(kwargs.keys()):
+            raise MlflowException(
+                "Missing keys in Array JSON. Expected to find keys `items` and `type`"
+            )
+        if kwargs["type"] != "array":
+            raise MlflowException("Type mismatch, Array expects `array` as the type")
+        if isinstance(kwargs["items"], str):
+            return cls(dtype=kwargs["items"])
+        if not isinstance(kwargs["items"], dict):
+            raise MlflowException("Expected items to be a dictionary of Object JSON")
+        return cls(Object.from_json_dict(**kwargs["items"]))
+
+
 class ColSpec:
     """
     Specification of name and type of a single column in a dataset.
     """
 
     def __init__(
         self,
-        type: Union[DataType, str],  # pylint: disable=redefined-builtin
+        type: Union[DataType, Array, Object, str],  # pylint: disable=redefined-builtin
         name: Optional[str] = None,
-        optional: bool = False,
+        optional: bool = None,
+        required: bool = None,  # TODO: update to required=True after deprecating optional
     ):
         self._name = name
-        self._optional = optional
+
+        if optional is not None:
+            if required is not None:
+                raise MlflowException(
+                    "Only one of `optional` and `required` can be specified."
+                    "`optional` is deprecated, please use `required` instead."
+                )
+            else:
+                _logger.warning(
+                    "`optional` is deprecated and will be removed in a future version "
+                    "of MLflow. Use `required` instead."
+                )
```

Comment:
/ai rewrite this using `warnings` module and `FutureWarning`. Explain why that's preferred.

Reply:
`_logger.warning` prints out the message every time we call it, while `FutureWarning` doesn't.